### PR TITLE
YG-412: UI - V2 >> Refactor VisibleColumnConfig on admin page.

### PR DESCRIPTION
**Pull request by pepcus-akshita**
# Related issues
- Refactored VisibleColumnConfig functionality on admin page.

# Proposed changes
- Added See All/See Less feature that adds/removes the History Details keys.
- Added Select All feature.
- Updated the config and separated as two different configs with current year details and previous year details.

### CODE REVIEW

File: src/v2/layouts/MemberDataGrid/MemberColumnConfig/DataGridColumnConfig.jsx [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+					</StyledCheckboxSelectContainer>
+					<div>
+						<StyledLink onClick={handleSeeAllLinkClick}>
+							{showAllCheckboxes ? SEE_LESS : SEE_ALL}
+						</StyledLink>
+					</div>
+					<div>
+						<SecondaryButton
+							text={CLOSE}
+							onClick={() => {
```
Comment (pepcus-mohit) [inline src/v2/layouts/MemberDataGrid/MemberColumnConfig/DataGridColumnConfig.jsx:82]:
```suggestion
							onClick={onClickClose}
```

File: src/v2/layouts/MemberDataGrid/MemberColumnConfig/DataGridColumnConfig.jsx [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	};
+
+	const handleSubmit = (formData) => {
+		onUpdateConfig(formData);
+		onClickClose();
+	};
+
+	const handleOnInputChange = (values) => {
+		const { selectAll } = values;
+		Object.keys(values).forEach((checkBoxInput) => {
```
Comment (pepcus-mohit) [inline src/v2/layouts/MemberDataGrid/MemberColumnConfig/DataGridColumnConfig.jsx:36]: I think this logic might be causing trouble in updating states.

File: src/v2/layouts/MemberDataGrid/MemberColumnConfig/DataGridColumnConfig.jsx [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+import PropTypes from 'prop-types';
+
+export const DataGridColumnConfig = ({ onClickClose, onUpdateConfig, initialConfig }) => {
+	const { strings } = useContext(StringsContext);
+	const { PLEASE_SELECT_COLUMNS_TEXT, SELECT_ALL, CLOSE, SAVE, SEE_ALL, SEE_LESS } = strings;
+	const [visibleCheckboxInputs, setVisibleCheckboxInputs] = useState([...currentYearGridColumns]);
+	const [showAllCheckboxes, setShowAllCheckboxes] = useState(false);
+
+	const updateVisibleCheckboxes = () => {
+		if (showAllCheckboxes === false) {
```
Comment (pepcus-mohit) [inline src/v2/layouts/MemberDataGrid/MemberColumnConfig/DataGridColumnConfig.jsx:18]:
```suggestion
		if (!showAllCheckboxes) {
```

File: src/v2/layouts/MemberDataGrid/MemberColumnConfig/DataGridColumnConfig.jsx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+				validate={handleOnInputChange}>
+				<Form>
+					<StyledCheckboxSelectContainer>
+						<CustomInput
+							key={SELECT_ALL}
+							type="checkbox"
+							name="selectAll"
+							label={SELECT_ALL}
+						/>
+						{visibleCheckboxInputs.map((data) => {
```
Comment (pepcus-mohit) [inline src/v2/layouts/MemberDataGrid/MemberColumnConfig/DataGridColumnConfig.jsx:62]:
```suggestion
						{visibleCheckboxInputs?.map((data) => {
```

File: src/v2/utils/dataGridUtils.js [hunk 5]
```diff
@@ -123,25 +129,41 @@ export const formatSelectedRow = (selectedRow, studentList) => {
...
+		key: 'edit',
+		label: 'Edit',
+		customComponent: EditMemberButton,
+	});
+	return { ...metaData, headerConfig };
+};
+
+export const formatGridColumnsForConfig = (gridColumns, defaultCheckedValue) => {
+	const formattedGridColumns = {};
+	gridColumns?.forEach((column) => {
```
Comment (pepcus-mohit) [inline src/v2/utils/dataGridUtils.js:164]: a map would be a better approach instead of using forEach

File: src/v2/utils/dataGridUtils.js [hunk 6]
```diff
@@ -123,25 +129,41 @@ export const formatSelectedRow = (selectedRow, studentList) => {
...
-	const editConfig = headerConfig.find((columnConfig) => columnConfig.key === 'edit') || {};
-	if ((isSelectAll || !isSelectAll) && updatedColumnList.length) {
-		updatedHeaderConfig = [editConfig, ...updatedColumnList];
-	} else if (isSelectAll) {
-		updatedHeaderConfig = headerConfig;
+export const updateStudentGridColumns = (checkedColumnList) => {
+	const headerConfig = [];
+	for (const key in checkedColumnList) {
+		if (checkedColumnList[key] === true) {
+			[...currentYearGridColumns, ...previousYearGridColumn].forEach((column) => {
```
Comment (pepcus-mohit) [inline src/v2/utils/dataGridUtils.js:136]:
```suggestion
			[...currentYearGridColumns, ...previousYearGridColumn]?.forEach((column) => {
```

File: src/v2/utils/dataGridUtils.js [hunk 7]
```diff
@@ -123,25 +129,41 @@ export const formatSelectedRow = (selectedRow, studentList) => {
...
-		updatedVisibleColumList.filter((columnConfig) => !columnConfig.isVisible) || [];
-	const editConfig = headerConfig.find((columnConfig) => columnConfig.key === 'edit') || {};
-	if ((isSelectAll || !isSelectAll) && updatedColumnList.length) {
-		updatedHeaderConfig = [editConfig, ...updatedColumnList];
-	} else if (isSelectAll) {
-		updatedHeaderConfig = headerConfig;
+export const updateStudentGridColumns = (checkedColumnList) => {
+	const headerConfig = [];
+	for (const key in checkedColumnList) {
+		if (checkedColumnList[key] === true) {
```
Comment (pepcus-mohit) [inline src/v2/utils/dataGridUtils.js:135]:
```suggestion
		if (checkedColumnList[key]) {
```

File: src/v2/layouts/MemberDataGrid/MemberSearchConfig/MemberSearchTextBox.jsx
Comment (pepcus-mohit) [inline src/v2/layouts/MemberDataGrid/MemberSearchConfig/MemberSearchTextBox.jsx:1]: If we keep code uncommented? Will that trigger any error?

**Review by pepcus-mohit — CHANGES_REQUESTED** [review]

File: src/v2/layouts/MemberDataGrid/MemberSearchConfig/MemberSearchTextBox.jsx
Comment (pepcus-akshita) [reply src/v2/layouts/MemberDataGrid/MemberSearchConfig/MemberSearchTextBox.jsx:1]: Yes, Formik's error will occur. As CheckBox component is already updated with new code and in this search option we are using Checkbox component only.

File: src/v2/utils/dataGridUtils.js (hunk 5 above)
Comment (pepcus-akshita) [reply src/v2/utils/dataGridUtils.js:164]: We are returning an object hence used forEach method.

**Review by pepcus-mohit — APPROVED** [review]

**Pull request merged**